### ISSUE TITLE
Opera Mini

### ISSUE THREAD

**Issue by Lisa-Williams** [sapo/Ink#425]
Hi,

Is Opera Mini 7.0.1.32694 supported by Ink?
If that specific version (ancient!!!) is not supported, but Opera Mini is supported, from which version up?

Cheers

**Comment by pedrocorreia** [comment]
Hi Lisa. For what i could tell from installing opera mini in my Android phone, a few hours ago,

**Comment by Lisa-Williams** [comment]
Hi @pedrocorreia thanks for looking into this.

When you say ``` "...the HTML/CSS portions of Ink seem to play well with it. ..."``` , does that include, *for example*, a working off-canvas menu (open and close *without* CSS animation of course)?
I realize that Opera Mini's way of handling JavaScript could still spoil this.
(Mainly server side JavaScript and only very limited client side JavaScript.)

*Some background:*
The reason I opened this issue is because I'm looking for a framework with *relatively* good Opera support (desktop and mobile) *to replace* another framework.
See [this issue](https://github.com/zurb/foundation/issues/3864) at their issue tracker.
But to be fair to them, ```ever since their last comment in that issue```, they do not mention Opera anymore in [their browser compatibility list](http://foundation.zurb.com/docs/compatibility.html).

I had already started coding a JavaScript to detect Opera Mini and fix the off-canvas menu in the other framework. But then I realized the probability of having to do that for (many) more components of that framework.
So I decided to look for another framework with better Opera support. Because that will be faster and less of a maintenance burden.

**Comment by pedrocorreia** [comment]
Well, out of the box the off canvas menu seems to be broken on Opera Mini. I suspect that it may be a JS issue. I'll bring @fabiosantoscode to the discussion as he is our lead JS guy. Maybe he can shed additional light on this matter.

**Comment by fabiosantoscode** [comment]
Hey @Lisa-Williams!

You've made me download a browser I'd never used before and learn how different and special it is :) Thanks!

That said, we don't really support it.

To put it bluntly, Opera mini is just too different a browser. We've never supported it and it's not in our plans to do so. While it certainly plays along well with Ink's CSS, it doesn't with our javascript which relies a lot on bubbling, scroll+resize events, transitions, touch events, and pretty much assumes that all the processing is happening on one device, which is not the case with Opera mini.

Cheers!

**Comment by pedrocorreia** [comment]
I must confess that we never gave Opera Mini much consideration and pretty much assumed that it was a scaled down version of the "regular" browser. But as Fábio says it's a completely different beast. 

From the Opera developer blog (https://dev.opera.com/articles/opera-mini-and-javascript/):

Much of the reason for Mini’s popularity is that it is capable of rendering complex web pages, even on low-spec devices. It does this by using a proxy-based architecture (see Figure 1). Requests from the user’s handset pass through the carrier’s internet gateway on their way to Opera’s transcoding servers. These servers then forward the request to the server.

Opera Mini requests pass through Opera’s servers

The server sends the response back as normal — when this is received by the Opera transcoding servers, they parse the markup and styles, execute the JavaScript, and transcode the data into Opera Binary Markup Language (OBML). This OBML data is progressively loaded by Opera Mini on the user’s device. Think of it as an interactive snapshot of a document’s state, which is similar in form to a PDF.

So unfortunately, as Fábio said, i don't see it becoming supported in a near future, if at all.

Best Regards.

**Comment by Lisa-Williams** [comment]
@fabiosantoscode and @pedrocorreia I understand it completely.
Because it's difficult to code around the limitations and time consuming.

I tried but never even got to run my Javacript once, because I gave up. Because I need to know what changes trigger a new trip to Opera's servers (after the page has been rendered) so that my open/close JavaScript gets processed by their servers.
Apparently adding and removing classes from JavaScript (after a click) works on the client side, and doesn't require a new trip to Opera's servers to run the JavaScript code.
*Aside: I can make a gist of what I have so far (Opera browser detection Mini, 12.x and 15+, and untested opening and closing the F5 off-canvas for Mini only).*

@pedrocorreia The page that you linked and [this one](https://dev.opera.com/articles/opera-mini-content-authoring-guidelines/#javascript) were what I used while attempting to code a JavaScript fix for the other framework's off-canvas.

[Another challenge](http://stackoverflow.com/questions/8307020/for-opera-mini-detect-if-mobile-view-is-on-or-off-using-javascript) I bumped in, is that Opera Mini has a [mobile view](http://www.opera.com/help/mini/guide), where Opera Mini takes care of re-ordering and stacking of page blocks: ```Mobile View: Disable page overview and zoom and display webpages to suit a small screen.```
In mobile view mode it made a complete mess of an other simple responsive test page.

Thanks to both of you for putting in the effort to try and help out. :)
